### PR TITLE
Rename real to scalar_t.

### PR DESCRIPTION
**Pull request by ezyang**
This is necessary to allow us to use the complex header
which defines real (and is very sad if real is macro'ed).

We should also fix accreal, ureal, Real and REAL, but
only 'real' is the real blocker.

```
codemod -d aten/src/TH --extensions c,cc,cpp,cu,cuh,h,TARGETS,py,hpp '\breal\b' scalar_t
codemod -d aten/src/THC --extensions c,cc,cpp,cu,cuh,h,TARGETS,py,hpp '\breal\b' scalar_t
codemod -d aten/src/THNN --extensions c,cc,cpp,cu,cuh,h,TARGETS,py,hpp '\breal\b' scalar_t
codemod -d aten/src/THCUNN --extensions c,cc,cpp,cu,cuh,h,TARGETS,py,hpp '\breal\b' scalar_t
```

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

### CODE REVIEW

File: aten/src/TH/generic/THTensorApply.hpp [hunk 1]
```diff
@@ -110,12 +110,12 @@
 }
 
 // Used for `scatter` and `scatterAdd`
-// Assumes TENSOR1 is real
+// Assumes TENSOR1 is scalar_t
 //         TENSOR2 is src
 //         TENSOR3 is index
 // Tests:
 //   1. index->size(d) <= src->size(d) for all d
-//   2. index->size(d) <= real->size(d) for all d != dim
```
Comment (SsnL) [inline aten/src/TH/generic/THTensorApply.hpp]: LOL `codemod` isn't smart enough. Can you revert the changes in this comment?

File: aten/src/TH/generic/THTensorLapack.cpp [hunk 2]
```diff
@@ -702,7 +702,7 @@ void THTensor_(potri)(THTensor *ra_, THTensor *a, const char *uplo)
 }
 
 /*
- Computes the Cholesky factorization with complete pivoting of a real symmetric
```
Comment (SsnL) [inline aten/src/TH/generic/THTensorLapack.cpp]: revert here too. is there an option in `codemod` that makes it ignore comments?

File: aten/src/TH/generic/THTensorLapack.cpp [hunk 3]
```diff
@@ -820,17 +820,17 @@ void THTensor_(geqrf)(THTensor *ra_, THTensor *rtau_, THTensor *a)
...
-  real wkopt = 0;
-  THLapack_(geqrf)(m, n, ra__->data<real>(), lda,
-                   rtau_->data<real>(),
+  scalar_t wkopt = 0;
+  THLapack_(geqrf)(m, n, ra__->data<scalar_t>(), lda,
+                   rtau_->data<scalar_t>(),
                    &wkopt, -1, &info);
 
-  /* Allocate the workspace and call LAPACK to do the real work. */
+  /* Allocate the workspace and call LAPACK to do the scalar_t work. */
```
Comment (SsnL) [inline aten/src/TH/generic/THTensorLapack.cpp]: here too

File: aten/src/TH/generic/THTensorLapack.cpp [hunk 4]
```diff
@@ -872,17 +872,17 @@ void THTensor_(orgqr)(THTensor *ra_, THTensor *a, THTensor *tau)
...
-  real wkopt = 0;
-  THLapack_(orgqr)(m, k, k, ra__->data<real>(), lda,
-                   tau->data<real>(),
+  scalar_t wkopt = 0;
+  THLapack_(orgqr)(m, k, k, ra__->data<scalar_t>(), lda,
+                   tau->data<scalar_t>(),
                    &wkopt, -1, &info);
 
-  /* Allocate the workspace and call LAPACK to do the real work. */
+  /* Allocate the workspace and call LAPACK to do the scalar_t work. */
```
Comment (SsnL) [inline aten/src/TH/generic/THTensorLapack.cpp]: here too

File: aten/src/THC/THCReduce.cuh [hunk 5]
```diff
@@ -359,7 +359,7 @@ inline dim3 getContigReduceBlock(ptrdiff_t numSlices, int64_t reductionSize) {
   // If the number of slices is low but the reduction dimension size
   // is high, then we should increase block size for greater parallelism.
   // Aim for at least 32 warps per SM (assume 15 SMs; don't bother
-  // inquiring the real number for now).
+  // inquiring the scalar_t number for now).
```
Comment (SsnL) [inline aten/src/THC/THCReduce.cuh]: here too: the real number

File: aten/src/THC/generic/THCTensorSort.cu [hunk 6]
```diff
@@ -254,7 +254,7 @@ void THCTensor_(sortViaThrust)(THCState* state,
     indexIter, indexIter + totalElements, keyIter,
     SliceComp(sliceSize));
 
-  // Translate the global integer 0-based index to a per-slice real
+  // Translate the global integer 0-based index to a per-slice scalar_t
```
Comment (SsnL) [inline aten/src/THC/generic/THCTensorSort.cu]: here too

File: aten/src/THNN/generic/Col2Im.c [hunk 7]
```diff
@@ -28,7 +28,7 @@
 // can slide the stencil over to the right once and down once).  But that is all
 // out-of-bounds if you're computing backwards for a 1x1 input.
 //
-// "Now Edward," you might say, "the real problem is that you set output_padding
+// "Now Edward," you might say, "the scalar_t problem is that you set output_padding
```
Comment (SsnL) [inline aten/src/THNN/generic/Col2Im.c]: the real problem :P

**Review by ssnl — COMMENTED** [review]
Changes mostly looking good, except for a couple places where `codemod` shouldn't change things (e.g., in some comments). I commented at places I noticed, but I'm not sure if I didn't miss something. 

A <del>real</del> scalar_t concern I have is `THCNumerics.cuh` where we define 
```cpp
template <typename scalar_t>
static inline __host__ __device__ scalar_t powi(scalar_t a, scalar_t b);
```

gcc7 complains when I do 
```cpp
<source>:7:23: error: expected nested-name-specifier before 'double'

     template<typename double>

                       ^~~~~~
```

It seems that it doesn't cause problem currently. But I wonder if this will be problematic if in future someone include this header in some file.

**Comment by SsnL** [comment]
Expanding all files gave me the most laggy chrome behavior I've ever experienced.

File: aten/src/TH/generic/THTensorLapack.cpp (hunk 2 above)
Comment (SsnL) [reply aten/src/TH/generic/THTensorLapack.cpp]: I checked `codemod` doc and there doesn't seem to be such an option

**Comment by ezyang** [comment]
Here's how I'm addressing the comments:

* I grepped for `ag "typename scalar_t"` in the TH directories, and manually fixed all occurrences that popped up by replacing `scalar_t` with `T`
* Fixed the referenced comment issues
* Searched for "//.+scalar_t" to see if I missed any inline comment issues
* Grepped the diff for /*, which actually did bring up some more comment issues

Thinking about the template parameter problem, I'm actually not sure how wise it is to rename this as `scalar_t`, because that's our canonical name for our templated methods, which means that if we include any of the `template<typename scalar_t>` headers from inside generic TH, we'll twing the error. I guess this shouldn't actually happen.

**Comment by ezyang** [comment]
Needs fbcode changes.

**Comment by SsnL** [comment]
New commits looks good! Let me know if you have the fbcode changes ready and I'll review.
